README
NAME
    Template::Plugin::Tooltip - Template Toolkit plugin for
    HTML::Tooltip::JavaScript

SYNOPSIS
      Load the tooltip generator.
      Params are passed straight to to HTML::Tooltip::JavaScript->new
      [% USE Tooltip( 'javascript_dir', '../../../static/js/', 'options',
            bordercolor => "#000000",
            title       => '',
            ) %]
  
      Add a tooltip to a link
      <a href="foo" [% tooltip(
          html_tooltip_content,
          param, value,
          param, value,
          %]>Link content</a>

DESCRIPTION
    Template::Plugin::Tooltip is a Template Toolkit hook to the marvelous
    HTML::Tooltip::Javascript module.

    The first version was written in 30 minutes following the talk on
    HTML::Tooltip::Javascript at the Open Source Developers Conference, just
    after its initial release.

  API Overview
    This module is very much just a thin layer over the top of the
    HTML::Tooltip::JavaScript API, and you should probably go read and
    understand its API before using this module.

    To summarise VERY briefly, when you load in the plugin, the params go as
    params to the H::T::Javascript "new" constructor, with the new Tooltip
    object stored internally. When you create a tooltip, the HTML content
    and parameters are passed directly to H::T::Javascript "tooltip" method.

  Loading the Tooltip Generator
    To load the tooltip generator with the default options, you can simple
    do the following.

      [% USE Tooltip %]

    In the same way you pass params to the HTML::Tooltip::Javascript
    constructor, you can also pass params when loading in the Tooltip
    plugin.

      [% USE Tooltip('param', 'value') %]

  Using the Tooltip Generator
    HTML::Tooltip::Javascript provides one simple method through which you
    generate all the different tooltips.

    In Template::Plugin::Tooltip, you just use the loaded plugin directly.

      [% tooltip( 'This is my plain tooltip' ) %]

    This only generates the Javascript tag properties, so this needs to be
    used within a tag, like an anchor tag.

      <a href="#item" [% tooltip('A tooltip') %]>item</a>

    The first param is literal HTML content, and you can provide any
    additional parameters you want for the tooltip, as you would do create
    the tooltip directly.

      <a href="#item" [% tooltip('A tooltip', 'bgcolor', 'pink') %]>item</a>

  Initialising the Tooltip Library
    The one additional step you will need to do is load in the tooltip
    JavaScript library that drives the whole thing.

    To do this, simple add the following to the end of the page, or to the
    <head> section of your HTML document.

      [% tooltip() %]

  Use as a Filter
    TO BE COMPLETED

SUPPORT
    Bugs should be submitted via the CPAN bug tracker, located at

    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Template-Plugin-Tooltip>

    For other issues, or commercial enhancement or support, contact the
    author..

AUTHOR
    Adam Kennedy , <http://ali.as/>, cpan@ali.as

ACKOWLEDGEMENTS
    Thank you to Phase N Australia (<http://phase-n.com/>) for permitting
    the open sourcing and release of this distribution as a spin-off from a
    commercial project.

COPYRIGHT
    Copyright (c) 2004-2005 Adam Kennedy. All rights reserved. This program
    is free software; you can redistribute it and/or modify it under the
    same terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.



Changes
Revision history for Perl extension Template::Plugin::Tooltip

0.05 Fri Dec 30 2005
	- More POD corrections (sigh)
	- Updated Module::Install

0.04 Mon May 23 2005
	- Minor POD corrections

0.03 Sun Apr 17 2005
	- Minor POD corrections

0.02 Tue Feb 15 2005
	- Removing braindead Build.PL

0.01 Wed Dec  1 2004
	- Created the initial version of the file


inc/Module/Install.pm
#line 1 "/home/adam/CPAN/packaged/Template-Plugin-Tooltip/inc/Module/Install.pm - /usr/local/share/perl/5.8.4/Module/Install.pm"
package Module::Install;
use 5.004;

$VERSION = '0.50'; # Don't forget to update Module::Install::Admin too!

die << "." unless $INC{join('/', inc => split(/::/, __PACKAGE__)).'.pm'};
Please invoke ${\__PACKAGE__} with:

    use inc::${\__PACKAGE__};

not:

    use ${\__PACKAGE__};

.

use strict 'vars';
use Cwd qw(cwd abs_path);
use FindBin;
use File::Find ();
use File::Path ();

@inc::Module::Install::ISA = 'Module::Install';
*inc::Module::Install::VERSION = *VERSION;

sub autoload {
    my $self   = shift;
    my $caller = $self->_caller;

    my $cwd = cwd();
    my $sym = "$caller\::AUTOLOAD";

    $sym->{$cwd} = sub {
        my $pwd = cwd();
        if (my $code = $sym->{$pwd}) {
            goto &$code unless $cwd eq $pwd; # delegate back to parent dirs
        }
        $$sym =~ /([^:]+)$/ or die "Cannot autoload $caller - $sym";
        unshift @_, ($self, $1);
        goto &{$self->can('call')} unless uc($1) eq $1;
    };
}

sub import {
    my $class = shift;
    my $self = $class->new(@_);

    if (not -f $self->{file}) {
        require "$self->{path}/$self->{dispatch}.pm";
        File::Path::mkpath("$self->{prefix}/$self->{author}");
        $self->{admin} = 
          "$self->{name}::$self->{dispatch}"->new(_top => $self);
        $self->{admin}->init;
        @_ = ($class, _self => $self);
        goto &{"$self->{name}::import"};
    }

    *{$self->_caller . "::AUTOLOAD"} = $self->autoload;
    $self->preload;

    # Unregister loader and worker packages so subdirs can use them again
    delete $INC{"$self->{file}"};
    delete $INC{"$self->{path}.pm"};
}

sub preload {
    my ($self) = @_;

    $self->load_extensions(
        "$self->{prefix}/$self->{path}", $self
    ) unless $self->{extensions};

    my @exts = @{$self->{extensions}};

    unless (@exts) {
        my $admin = $self->{admin};
        @exts = $admin->load_all_extensions;
    }

    my %seen_method;
    foreach my $obj (@exts) {
        while (my ($method, $glob) = each %{ref($obj) . '::'}) {
            next unless defined *{$glob}{CODE};
            next if $method =~ /^_/;
            next if $method eq uc($method);
            $seen_method{$method}++;
        }
    }

    my $caller = $self->_caller;
    foreach my $name (sort keys %seen_method) {
        *{"${caller}::$name"} = sub {
            ${"${caller}::AUTOLOAD"} = "${caller}::$name";
            goto &{"${caller}::AUTOLOAD"};
        };
    }
}

sub new {
    my ($class, %args) = @_;

    # ignore the prefix on extension modules built from top level.
    my $base_path = abs_path($FindBin::Bin);
    delete $args{prefix} unless abs_path(cwd()) eq $base_path;

    return $args{_self} if $args{_self};

    $args{dispatch} ||= 'Admin';
    $args{prefix}   ||= 'inc';
    $args{author}   ||= '.author';
    $args{bundle}   ||= 'inc/BUNDLES';
    $args{base}     ||= $base_path;

    $class =~ s/^\Q$args{prefix}\E:://;
    $args{name}     ||= $class;
    $args{version}  ||= $class->VERSION;

    unless ($args{path}) {
        $args{path}  = $args{name};
        $args{path}  =~ s!::!/!g;
    }
    $args{file}     ||= "$args{base}/$args{prefix}/$args{path}.pm";

    bless(\%args, $class);
}

sub call {
    my $self   = shift;
    my $method = shift;
    my $obj    = $self->load($method) or return;

    unshift @_, $obj;
    goto &{$obj->can($method)};
}

sub load {
    my ($self, $method) = @_;

    $self->load_extensions(
        "$self->{prefix}/$self->{path}", $self
    ) unless $self->{extensions};

    foreach my $obj (@{$self->{extensions}}) {
        return $obj if $obj->can($method);
    }

    my $admin = $self->{admin} or die << "END";
The '$method' method does not exist in the '$self->{prefix}' path!
Please remove the '$self->{prefix}' directory and run $0 again to load it.
END

    my $obj = $admin->load($method, 1);
    push @{$self->{extensions}}, $obj;

    $obj;
}

sub load_extensions {
    my ($self, $path, $top_obj) = @_;

    unshift @INC, $self->{prefix}
        unless grep { $_ eq $self->{prefix} } @INC;

    local @INC = ($path, @INC);
    foreach my $rv ($self->find_extensions($path)) {
        my ($file, $pkg) = @{$rv};
        next if $self->{pathnames}{$pkg};

        local $@;
        my $new = eval { require $file; $pkg->can('new') };
        if (!$new) { warn $@ if $@; next; }
        $self->{pathnames}{$pkg} = delete $INC{$file};
        push @{$self->{extensions}}, &{$new}($pkg, _top => $top_obj );
    }

    $self->{extensions} ||= [];
}

sub find_extensions {
    my ($self, $path) = @_;
    my @found;

    File::Find::find(sub {
        my $file = $File::Find::name;
        return unless $file =~ m!^\Q$path\E/(.+)\.pm\Z!is;
        return if $1 eq $self->{dispatch};

        $file = "$self->{path}/$1.pm";
        my $pkg = "$self->{name}::$1"; $pkg =~ s!/!::!g;
        push @found, [$file, $pkg];
    }, $path) if -d $path;

    @found;
}

sub _caller {
    my $depth = 0;
    my $caller = caller($depth);

    while ($caller eq __PACKAGE__) {
        $depth++;
        $caller = caller($depth);
    }

    $caller;
}

1;


lib/Template/Plugin/Tooltip.pm
package Template::Plugin::Tooltip;

=pod

=head1 NAME

Template::Plugin::Tooltip - Template Toolkit plugin for HTML::Tooltip::JavaScript

=head1 SYNOPSIS

  Load the tooltip generator.
  Params are passed straight to to HTML::Tooltip::JavaScript->new
  [% USE Tooltip( 'javascript_dir', '../../../static/js/', 'options',
	bordercolor => "#000000",
	title       => '',
	) %]
  
  Add a tooltip to a link
  <a href="foo" [% tooltip(
      html_tooltip_content,
      param, value,
      param, value,
      %]>Link content</a>

=head1 DESCRIPTION

Template::Plugin::Tooltip is a Template Toolkit hook to the marvelous
L<HTML::Tooltip::Javascript> module.

The first version was written in 30 minutes following the talk on
HTML::Tooltip::Javascript at the Open Source Developers Conference,
just after its initial release.

=head2 API Overview

This module is very much just a thin layer over the top of the
HTML::Tooltip::JavaScript API, and you should probably go read and
understand its API before using this module.

To summarise VERY briefly, when you load in the plugin, the params go as
params to the H::T::Javascript C<new> constructor, with the new Tooltip
object stored internally. When you create a tooltip, the HTML content and
parameters are passed directly to H::T::Javascript C<tooltip> method.

=head2 Loading the Tooltip Generator

To load the tooltip generator with the default options, you can simple do
the following.

  [% USE Tooltip %]

In the same way you pass params to the HTML::Tooltip::Javascript
constructor, you can also pass params when loading in the Tooltip plugin.

  [% USE Tooltip('param', 'value') %]

=head2 Using the Tooltip Generator

L<HTML::Tooltip::Javascript> provides one simple method through which
you generate all the different tooltips.

In Template::Plugin::Tooltip, you just use the loaded plugin directly.

  [% tooltip( 'This is my plain tooltip' ) %]

This only generates the Javascript tag properties, so this needs to be used
within a tag, like an anchor tag.

  <a href="#item" [% tooltip('A tooltip') %]>item</a>

The first param is literal HTML content, and you can provide any additional
parameters you want for the tooltip, as you would do create the tooltip
directly.

  <a href="#item" [% tooltip('A tooltip', 'bgcolor', 'pink') %]>item</a>

=head2 Initialising the Tooltip Library

The one additional step you will need to do is load in the tooltip
JavaScript library that drives the whole thing.

To do this, simple add the following to the end of the page, or to the
E<lt>headE<gt> section of your HTML document.

  [% tooltip() %]

=head2 Use as a Filter

TO BE COMPLETED

=cut

use strict;
use base 'Template::Plugin';
use HTML::Tooltip::Javascript ();

use vars qw{$VERSION};
BEGIN {
	$VERSION = '0.05';
}





#####################################################################
# Constructor

sub new {
	my $class   = ref $_[0] ? ref shift : shift;
	my $Context = UNIVERSAL::isa(ref $_[0], 'Template::Context') ? shift
		: $class->error('Tooltip constructor called without Template::Context');

	# Create a new ::Tooltip object, passing on params
	my $Tooltip = HTML::Tooltip::Javascript->new( @_ )
		or $class->error('Bad params to Tooltip generator constructor');

	# Create our TT interface object
	my $self = bless {
		Tooltip => $Tooltip,
		}, $class;

	# Wrap in the closure subroutine
	sub { defined $_[0] ? $self->tooltip(@_) : $self->head };
}

sub tooltip {
	my $self = shift;

	# If we don't get any parameters, we want a direct handle to the
	# object so that we can call init.
	return $self unless @_;

	# Pass everything straight through to the HTML::Tooltip::JavaScript
	# object we created earlier.
	my ($html, %params) = @_;
	$self->{Tooltip}->tooltip( $html, \%params );
}

sub head {
	shift->at_end;
}

sub at_end {
	shift->{Tooltip}->at_end;
}

1;

=pod

=head1 SUPPORT

Bugs should be submitted via the CPAN bug tracker, located at

L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Template-Plugin-Tooltip>

For other issues, or commercial enhancement or support, contact the author..

=head1 AUTHOR

Adam Kennedy , L<http://ali.as/>, cpan@ali.as

=head1 ACKOWLEDGEMENTS

Thank you to Phase N Australia (L<http://phase-n.com/>) for permitting the
open sourcing and release of this distribution as a spin-off from a
commercial project.

=head1 COPYRIGHT

Copyright (c) 2004-2005 Adam Kennedy. All rights reserved.
This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=cut
